File: src/hexafe_plotstats/renderers/matplotlib/histogram.py
from __future__ import annotations

import numpy as np
from matplotlib.lines import Line2D
from matplotlib.patches import Rectangle
from matplotlib.ticker import PercentFormatter

from ...models.payloads import HistogramPayload
from ...models.render import RenderResult
from ...specs import ResolvedHistogramSpec, TableRow, TableSpec
from ...specs import histogram_payload_to_resolved_spec
from .._common import apply_resolved_layout, style_spec_limits

_BADGE_COLORS = {
    "quality_capable": ("#dcfce7", "#166534"),
    "quality_good": ("#dbeafe", "#1e3a8a"),
    "quality_marginal": ("#fef3c7", "#92400e"),
    "quality_risk": ("#fee2e2", "#991b1b"),
    "quality_unknown": ("#e5e7eb", "#374151"),
    "fit_quality_high": ("#dcfce7", "#166534"),
    "fit_quality_medium": ("#fef3c7", "#92400e"),
    "fit_quality_low": ("#fee2e2", "#991b1b"),
    "normality_normal": ("#dcfce7", "#166534"),
    "normality_not_normal": ("#fee2e2", "#991b1b"),
}


def render_histogram_matplotlib(payload: HistogramPayload) -> RenderResult:
    import matplotlib.pyplot as plt

    fig, ax = plt.subplots()
    resolved = histogram_payload_to_resolved_spec(payload)
    edges = np.asarray(payload.bin_edges, dtype=float)
    heights = np.asarray(payload.bin_values, dtype=float)
    y_mode = str(
        payload.metadata.get("histogram_y_mode")
        or payload.metadata.get("y_mode")
        or payload.metadata.get("normalization")
        or ""
    ).strip().lower()
    relative_frequency = y_mode in {"relative_percent", "frequency_percent", "percent"}

    if edges.size >= 2 and heights.size:
        widths = np.diff(edges)
        if relative_frequency:
            counts, _ = np.histogram(np.asarray(payload.values, dtype=float), bins=edges)
            total = float(np.sum(counts))
            if total > 0.0:
                heights = counts.astype(float) / total
        ax.bar(edges[:-1], heights, width=widths, align="edge", alpha=0.8, color="tab:blue")

    curve_y_scale = 1.0
    if relative_frequency and edges.size >= 2:
        widths = np.diff(edges)
        positive_widths = widths[np.isfinite(widths) & (widths > 0.0)]
        if positive_widths.size:
            curve_y_scale = float(np.median(positive_widths))
    for curve in resolved.curves:
        curve_y = np.asarray(curve.y, dtype=float) * curve_y_scale
        if curve.fill_to_baseline and curve.fill_alpha > 0.0:
            ax.fill_between(curve.x, curve_y, 0.0, color=curve.fill_color or curve.stroke, alpha=curve.fill_alpha, linewidth=0.0)
        ax.plot(
            curve.x,
            curve_y,
            color=curve.stroke,
            linewidth=curve.stroke_width,
            linestyle="--" if curve.dash else "-",
            alpha=curve.opacity,
        )

    style_spec_limits(ax, payload.spec_limits)
    if relative_frequency:
        ax.yaxis.set_major_formatter(PercentFormatter(xmax=1.0))
    axis_labels = payload.metadata.get("axis_labels") if isinstance(payload.metadata.get("axis_labels"), dict) else {}
    ax.set_ylabel(str(axis_labels.get("y") or payload.metadata.get("y_label") or ("density" if payload.density else "count")))
    ax.set_xlabel(str(axis_labels.get("x") or payload.metadata.get("x_label") or "value"))
    if payload.metadata.get("title"):
        ax.set_title(str(payload.metadata["title"]))
    apply_resolved_layout(fig, ax, resolved)
    canvas = resolved.canvas.size
    for line in resolved.annotation_lines:
        fig.add_artist(
            Line2D(
                [line.x0 / canvas.width, line.x1 / canvas.width],
                [1.0 - (line.y0 / canvas.height), 1.0 - (line.y1 / canvas.height)],
                color=line.stroke,
                linewidth=line.stroke_width,
                alpha=0.72,
                transform=fig.transFigure,
            )
        )
    _draw_resolved_table(fig, resolved)
    return RenderResult(fig=fig, ax=ax, metadata={"kind": "histogram", "warnings": payload.warnings})


def _draw_resolved_table(fig, spec: ResolvedHistogramSpec) -> None:
    table = spec.table
    if table is None:
        return

    canvas = spec.canvas.size
    if canvas.width <= 0.0 or canvas.height <= 0.0:
        return

    rect = table.rect
    table_ax = fig.add_axes(
        [
            rect.x / canvas.width,
            1.0 - ((rect.y + rect.height) / canvas.height),
            rect.width / canvas.width,
            rect.height / canvas.height,
        ]
    )
    table_ax.set_xlim(0.0, 1.0)
    table_ax.set_ylim(0.0, 1.0)
    table_ax.set_xticks([])
    table_ax.set_yticks([])
    table_ax.set_facecolor("#ffffff")
    for spine in table_ax.spines.values():
        spine.set_color("#d1d5db")
        spine.set_linewidth(1.0)

    y = 16.0
    if table.header:
        _draw_table_row(table_ax, table, y, table.header, weight="600")
        y += 16.0
        separator_y = 1.0 - ((y - 4.0) / rect.height)
        table_ax.axhline(separator_y, color="#e5e7eb", linewidth=1.0)

    for row in table.rows:
        if y > rect.height - 8.0:
            break
        if row.metadata.get("section_break_before"):
            separator_y = 1.0 - ((y - 7.0) / rect.height)
            table_ax.axhline(separator_y, color="#d1d5db", linewidth=1.0)
        _draw_table_row(table_ax, table, y, row)
        y += 18.0


def _draw_table_row(table_ax, table: TableSpec, y: float, row, *, weight: str = "normal") -> None:
    if table.rect.width <= 0.0 or table.rect.height <= 0.0:
        return
    cells = row.cells if isinstance(row, TableRow) else row
    metadata = row.metadata if isinstance(row, TableRow) else {}
    y_axis = 1.0 - (y / table.rect.height)
    label = str(cells[0].text) if len(cells) >= 1 else ""
    value = str(cells[1].text) if len(cells) >= 2 else ""
    text_color = "#111827"
    badge_palette = metadata.get("badge_palette")
    if isinstance(badge_palette, str) and badge_palette in _BADGE_COLORS:
        background, text_color = _BADGE_COLORS[badge_palette]
        row_height = min(16.0 / table.rect.height, 0.12)
        table_ax.add_patch(
            Rectangle(
                (3.0 / table.rect.width, y_axis - row_height * 0.5),
                1.0 - (6.0 / table.rect.width),
                row_height,
                transform=table_ax.transAxes,
                facecolor=background,
                edgecolor="none",
                alpha=0.94,
                zorder=0,
            )
        )
    table_ax.text(
        8.0 / table.rect.width,
        y_axis,
        label,
        ha="left",
        va="center",
        fontsize=10,
        fontweight=weight,
        color=text_color if badge_palette else "#374151",
    )
    table_ax.text(
        1.0 - (8.0 / table.rect.width),
        y_axis,
        value,
        ha="right",
        va="center",
        fontsize=10,
        fontweight=weight,
        color=text_color,
    )

File: native/hexafe-plotstats-native/src/charts/histogram.rs
use crate::spec::{ChartTransform, HistogramSpec, Point, Rect, RenderResult, TextSpec};
use crate::svg::{
    draw_axes, draw_line_spec, draw_text_spec, finish_chart, render_surface, RasterCanvas,
    RenderedChart, SvgDocument,
};

pub fn render(spec: &HistogramSpec) -> RenderResult<RenderedChart> {
    let (mut svg, mut raster, width, height) = render_surface(&spec.common)?;
    draw_axes(&mut svg, &mut raster, &spec.common);
    let transform = ChartTransform::from_common(&spec.common);

    for bar in &spec.bars {
        let p0 = transform.map(bar.x0, bar.y0);
        let p1 = transform.map(bar.x1, bar.y1);
        let rect = Rect {
            x: p0.x.min(p1.x),
            y: p0.y.min(p1.y),
            width: (p0.x - p1.x).abs().max(1.0),
            height: (p0.y - p1.y).abs().max(1.0),
        };
        svg.rect(rect, &bar.fill, &bar.stroke, bar.opacity, 0.8);
        raster.rect(rect, &bar.fill, &bar.stroke, bar.opacity, 0.8);
    }

    for curve in &spec.curves {
        let count = curve.x.len().min(curve.y.len());
        let points = (0..count)
            .map(|idx| {
                if curve.coordinate_space == "canvas" {
                    Point {
                        x: curve.x[idx],
                        y: curve.y[idx],
                    }
                } else {
                    transform.map(curve.x[idx], curve.y[idx])
                }
            })
            .collect::<Vec<_>>();
        svg.polyline(
            &points,
            &curve.stroke,
            curve.stroke_width,
            &curve.dash,
            curve.opacity,
        );
        raster.stroke_polyline(&points, &curve.stroke, curve.stroke_width, curve.opacity);
    }

    for line in &spec.spec_lines {
        draw_line_spec(&mut svg, &mut raster, transform, line);
    }
    if let Some(line) = &spec.mean_line {
        draw_line_spec(&mut svg, &mut raster, transform, line);
    }
    for annotation in &spec.annotations {
        draw_text_spec(&mut svg, transform, annotation);
    }

    draw_table(spec, &mut svg, &mut raster);

    finish_chart(svg, raster, width, height)
}

fn draw_table(spec: &HistogramSpec, svg: &mut SvgDocument, raster: &mut RasterCanvas) {
    let Some(table) = &spec.table else {
        if let Some(rect) = spec.table_rect {
            svg.rect(rect, "#ffffff", "#d1d5db", 1.0, 1.0);
            raster.rect(rect, "#ffffff", "#d1d5db", 1.0, 1.0);
        }
        return;
    };

    svg.rect(table.rect, "#ffffff", "#d1d5db", 1.0, 1.0);
    raster.rect(table.rect, "#ffffff", "#d1d5db", 1.0, 1.0);

    let mut y = table.rect.y + 16.0;
    if !table.header.is_empty() {
        draw_table_text(
            svg,
            table.rect,
            y,
            table
                .header
                .get(0)
                .map(|cell| cell.text.clone())
                .unwrap_or_default(),
            table
                .header
                .get(1)
                .map(|cell| cell.text.clone())
                .unwrap_or_default(),
            "600",
            "#111827",
        );
        y += 16.0;
        let separator_y = y - 4.0;
        svg.line(
            Point {
                x: table.rect.x,
                y: separator_y,
            },
            Point {
                x: table.rect.x + table.rect.width,
                y: separator_y,
            },
            "#e5e7eb",
            1.0,
            &[],
            1.0,
        );
        raster.line(
            Point {
                x: table.rect.x,
                y: separator_y,
            },
            Point {
                x: table.rect.x + table.rect.width,
                y: separator_y,
            },
            "#e5e7eb",
            1.0,
            1.0,
        );
    }

    for row in &table.rows {
        if row.section_break_before {
            let separator_y = y - 7.0;
            svg.line(
                Point {
                    x: table.rect.x,
                    y: separator_y,
                },
                Point {
                    x: table.rect.x + table.rect.width,
                    y: separator_y,
                },
                "#d1d5db",
                1.0,
                &[],
                1.0,
            );
            raster.line(
                Point {
                    x: table.rect.x,
                    y: separator_y,
                },
                Point {
                    x: table.rect.x + table.rect.width,
                    y: separator_y,
                },
                "#d1d5db",
                1.0,
                1.0,
            );
        }
        let value_color = if let Some(palette) = &row.badge_palette {
            let (background, text_color) = badge_colors(palette);
            let badge_rect = Rect {
                x: table.rect.x + 3.0,
                y: y - 8.0,
                width: (table.rect.width - 6.0).max(0.0),
                height: 16.0,
            };
            svg.rect(badge_rect, background, "none", 0.94, 0.0);
            raster.rect(badge_rect, background, "none", 0.94, 0.0);
            text_color
        } else {
            "#111827"
        };
        let label = row
            .cells
            .get(0)
            .map(|cell| cell.text.clone())
            .unwrap_or_default();
        let value = row
            .cells
            .get(1)
            .map(|cell| cell.text.clone())
            .unwrap_or_default();
        draw_table_text(svg, table.rect, y, label, value, "normal", value_color);
        y += 18.0;
        if y > table.rect.y + table.rect.height - 8.0 {
            break;
        }
    }
}

fn draw_table_text(
    svg: &mut SvgDocument,
    rect: Rect,
    y: f64,
    label: String,
    value: String,
    weight: &str,
    value_color: &str,
) {
    svg.text(&TextSpec {
        text: label,
        x: rect.x + 8.0,
        y,
        font_size: 10.0,
        fill: if value_color == "#111827" {
            "#374151".to_string()
        } else {
            value_color.to_string()
        },
        align: "left".to_string(),
        baseline: "middle".to_string(),
        weight: weight.to_string(),
        coordinate_space_x: "canvas".to_string(),
        coordinate_space_y: "canvas".to_string(),
    });
    svg.text(&TextSpec {
        text: value,
        x: rect.x + rect.width - 8.0,
        y,
        font_size: 10.0,
        fill: value_color.to_string(),
        align: "right".to_string(),
        baseline: "middle".to_string(),
        weight: weight.to_string(),
        coordinate_space_x: "canvas".to_string(),
        coordinate_space_y: "canvas".to_string(),
    });
}

fn badge_colors(palette: &str) -> (&'static str, &'static str) {
    match palette {
        "quality_capable" | "fit_quality_high" | "normality_normal" => ("#dcfce7", "#166534"),
        "quality_good" => ("#dbeafe", "#1e3a8a"),
        "quality_marginal" | "fit_quality_medium" => ("#fef3c7", "#92400e"),
        "quality_risk" | "fit_quality_low" | "normality_not_normal" => ("#fee2e2", "#991b1b"),
        _ => ("#e5e7eb", "#374151"),
    }
}
